InstructionSets/x86/DataPointerResolver.hpp

```diff
@@ -0,0 +1,200 @@
+//
+//  DataPointerResolver.hpp
+//  Clock Signal
+//
+//  Created by Thomas Harte on 24/02/2022.
+//  Copyright © 2022 Thomas Harte. All rights reserved.
+//
+
+#ifndef DataPointerResolver_hpp
+#define DataPointerResolver_hpp
+
+#include "Instruction.hpp"
+#include "Model.hpp"
+
+namespace InstructionSet {
+namespace x86 {
+
+/// Unlike source, describes only registers, and breaks
+/// them down by conventional name — so AL, AH, AX and EAX are all
+/// listed separately and uniquely, rather than being eAX+size or
+/// eSPorAH with a size of 1.
+enum class Register: uint8_t {
+	AL, AH, AX, EAX,
+	CL, CH, CX, ECX,
+	DL, DH, DX, EDX,
+	BL, BH, BX, EBX,
+	SP, ESP,
+	BP, EBP,
+	SI, ESI,
+	DI, EDI,
+	ES,
+	CS,
+	SS,
+	DS,
+	FS,
+	GS,
+	None
+};
+
+/// Reads from or writes to the source or target identified by a DataPointer, relying upon two user-supplied classes:
+///
+/// * a register bank; and
+/// * a memory pool.
+///
+/// The register bank should implement `template<typename DataT, Register> DataT read()` and `template<typename DataT, Register> void write(DataT)`.
+/// Those functions will be called only with registers and data types that are appropriate to the @c model.
+///
+/// The memory pool should implement `template<typename DataT> DataT read(Source segment, uint32_t address)` and
+/// `template<typename DataT> void write(Source segment, uint32_t address, DataT value)`.
+template <Model model, typename RegistersT, typename MemoryT> class DataPointerResolver {
+	public:
+		template <typename DataT> DataT read(
+			RegistersT &registers,
+			MemoryT &memory,
+			const Instruction<is_32bit(model)> &instruction,
+			DataPointer pointer,
+			decltype(RegistersT::eSP) memory_mask = ~0) {
+				DataT result;
+				access<true>(registers, memory, instruction, pointer, memory_mask, result);
+				return result;
+			}
+
+		template <typename DataT> void write(
+			RegistersT &registers,
+			MemoryT &memory,
+			const Instruction<is_32bit(model)> &instruction,
+			DataPointer pointer,
+			DataT value,
+			decltype(RegistersT::eSP) memory_mask = ~0) {
+				access<false>(registers, memory, instruction, pointer, memory_mask, value);
+			}
+
+	private:
+		template <bool is_write, typename DataT> DataT access(
+			RegistersT &registers,
+			MemoryT &memory,
+			const Instruction<is_32bit(model)> &instruction,
+			DataPointer pointer,
+			decltype(RegistersT::eSP) memory_mask,
+			DataT &value) {
+				const Source source = pointer.source();
+
+#define read_or_write(v, x, allow_write)	\
```

Comment:
Perhaps replace `allow_write` by `is_not_indirect` (or invert that name, the use and the given argument values) to make it a bit more clear why this argument is switching from a write to a read?